### PR TITLE
Feature/831 create request behavior

### PR DESCRIPTION
**Pull request by Greenscreen23**
### PR Checklist

Please make sure to fulfil the following conditions before marking this PR ready for review:

-   [x] If this PR adds or changes features or fixes bugs, this has been added to the changelog
-   [x] If this PR adds new actions or other ways to alter the state, [test scenarios](https://github.com/hpi-sam/digital-fuesim-manv-public-test-scenarios) have been added

### CODE REVIEW

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    requestTargetOptions$!: Observable<{
+        [key in RequestTargetOption]: string;
+    }>;
+
+    waitingForAnswer$!: Observable<boolean>;
+
+    nextTimeoutIn$!: Observable<number>;
+
+    initialRequestTargetOption$!: Observable<RequestTargetOption>;
```
Comment (lukasrad02) [inline frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts:49]: Why does the name of this observable include "initial"?

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+        const currentTime$ = this.store.select(selectCurrentTime);
+
+        this.nextTimeoutIn$ = combineLatest([
+            this.requestBehaviorState$,
+            activities$,
+            currentTime$,
+        ]).pipe(
+            map(([requestBehaviorState, activities, currentTime]) => {
+                if (!requestBehaviorState.delayEventActivityId) return -1;
```
Comment (lukasrad02) [inline frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts:113]: You could use `isWaitingForTimeout` here.

**Review by lukasrad02 — COMMENTED** [review]
I'm not yet done with all files, but as you're going to change something in the behavior, I'll wait for it.

Here are already some things I've noticed in the frontend.

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.html [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                        class="form-control form-control-sm d-inline-block no-validation"
+                    />
+                    <span class="input-group-text">Min</span>
+                </div>
+            </div>
+        </div>
+        <div class="d-flex row">
+            <div class="col-sm">
+                <span
+                    >Dauer, nachdem ein nicht eingelöstes Versprechen nach
```
[the remark this answers is not in the document]
Comment (Greenscreen23) [reply frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.html]: Done in 103c3d3ddcd06afd35c9ce1aa574a9bf3fa3771e

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.html [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                nextTimeoutIn.time === -1
+            "
+            class="text-muted"
+            >Es gibt aktuell keine Fahrzeuganfragen in dieser Region</span
+        >
+        <span *ngIf="(waitingForAnswer$ | async) === true" class="text-muted"
+            >Wartet auf die Antwort einer Anfrage von</span
+        >
+        <span *ngIf="nextTimeoutIn.time !== -1" class="text-muted"
+            >Nächste Anfrage potenziell in
```
[the remark this answers is not in the document]
Comment (Greenscreen23) [reply frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.html]: Done in 103c3d3ddcd06afd35c9ce1aa574a9bf3fa3771e

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts (hunk 1 above)
Comment (Greenscreen23) [reply frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts]: I've changed it in 103c3d3ddcd06afd35c9ce1aa574a9bf3fa3771e

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts (hunk 2 above)
Comment (Greenscreen23) [reply frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.ts]: I've completely removed the deprecation in 103c3d3ddcd06afd35c9ce1aa574a9bf3fa3771e

File: frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.html
[the remark this answers is not in the document]
Comment (Greenscreen23) [reply frontend/src/app/pages/exercises/exercise/shared/simulated-region-overview/tabs/behavior-tab/behaviors/request-vehicles/simulated-region-overview-behavior-request-vehicles.component.html:1]: How do you like 395db4e2625ab2e324348f29dbe42ba4b6da1d68?

File: shared/src/simulation/behaviors/request.spec.ts [hunk 5]
```diff
@@ -0,0 +1,853 @@
...
+                            const promise = promisedResources.at(-1)!;
+                            expect(promise.resource).toEqual(
+                                VehicleResource.create({ KTW: 1 })
+                            );
+                        });
+
+                        if (
+                            state === 'waiting' &&
+                            event === 'vehiclesSendEventForAnswerKey'
+                        ) {
```
Comment (lukasrad02) [inline shared/src/simulation/behaviors/request.spec.ts:570]: I find it extremely hard to understand the tests with the large amount of conditionals in the test logic.

As you've already created functions that initialize the state I think it should be possible to have less conditionals without causing to much duplicated code.

File: shared/src/simulation/behaviors/request.spec.ts [hunk 6]
```diff
@@ -0,0 +1,853 @@
...
+const withoutVehiclesRequired: Set<keyof typeof addRequestsAndPromises> =
+    new Set([
+        'withoutRequestsAndPromises',
+        'withPromises',
+        'withRequestsAndEnoughPromises',
+        'withPromiseOfOtherType',
+        'withPromisesOfMultipleTypes',
+    ]);
+
+const oldTime = currentTime - 100;
```
[the remark this answers is not in the document]
Comment (Greenscreen23) [reply shared/src/simulation/behaviors/request.spec.ts]: I've done this in c7e2b0a2965f6b9f1b61752828557cb40f374f11

**Review by lukasrad02 — APPROVED** [review]
I did not review the new tests in detail, but since it was mainly a re-structuring, I trust you with this.

File: shared/src/state-migrations/26-add-keys-to-radiograms-and-events.ts
[the remark this answers is not in the document]
Comment (lukasrad02) [reply shared/src/state-migrations/26-add-keys-to-radiograms-and-events.ts:1]: I think no comment from you means that you do not plan to remove this migration. So I close this thread to unblock merging

[the remark this answers is not in the document]
Comment (Greenscreen23) [reply shared/src/state-migrations/26-add-keys-to-radiograms-and-events.ts:1]: I've reverted the migration

**Review by lukasrad02 — APPROVED** [review]
The migration tests are only failing because of the scenarios added by Nils PR, that is also currently open. So you may bypass the branch protection.

I've created #861 to investigate solutions for this.

**Pull request merged**